### PR TITLE
PIC-1070 Create PUT Preference Endpoint

### PR DESCRIPTION
**Pull request by theDustRoom**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/controller/CourtsController.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+@RestController
+class CourtsController {
+
+  @Autowired
+  lateinit var repository: UsersRepository
+
+  @ResponseStatus(value = HttpStatus.CREATED)
+  @PutMapping("/users/{uuid}/preferences/courts")
+  fun updatePreferences(@PathVariable uuid: UUID, @RequestBody user: Users): List<String>? =
```
Comment (chris-faulkner-mojdt) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/controller/CourtsController.kt:22]: Think we should ensure that the response is never a null  but an empty list of strings. If you remove the "?" isn't that saying the response will never be null ?

Comment (theDustRoom) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/controller/CourtsController.kt:22]: Returns: the saved entities; will never be null. The returned Iterable will have the same size as the Iterable passed as an argument.

Comment (theDustRoom) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/controller/CourtsController.kt:22]: Sorry, I misunderstood the question. Yes, I'll change it.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/jpa/entity/Users.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import org.hibernate.annotations.Type
+import org.hibernate.annotations.TypeDef
+import java.util.UUID
+import javax.persistence.Column
+import javax.persistence.Entity
+import javax.persistence.Id
+
+@Entity
+@TypeDef(name = "list-array", typeClass = ListArrayType::class)
+data class Users(
```
Comment (chris-faulkner-mojdt) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/jpa/entity/Users.kt:13]: not seen this done in Kotlin before for JPA.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/jpa/entity/Users.kt [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  @Id val id: UUID? = null,
+  @Type(type = "list-array") @Column(name = "courts", columnDefinition = "text[]")
+  var courts: List<String>? = null
```
Comment (chris-faulkner-mojdt) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/jpa/entity/Users.kt]: As discussed - need to make a general key/value pair modelling

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/controller/CourtsController.kt [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import java.util.UUID
+
+@RestController
+class CourtsController {
+
+  @Autowired
+  lateinit var repository: UsersRepository
+
+  @ResponseStatus(value = HttpStatus.CREATED)
+  @PutMapping("/users/{uuid}/preferences/courts")
```
Comment (chris-faulkner-mojdt) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/controller/CourtsController.kt:21]: Change the mapping to remove the "courts"

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/controller/CourtsController.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import uk.gov.justice.digital.hmpps.hmppsuserpreferences.jpa.repository.UsersRepository
+import java.util.UUID
+
+@RestController
+class CourtsController {
+
+  @Autowired
+  lateinit var repository: UsersRepository
+
+  @ResponseStatus(value = HttpStatus.CREATED)
```
Comment (chris-faulkner-mojdt) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/controller/CourtsController.kt:20]: Does the DPS project template bring in the Swagger components ? We could do with swagger annotations.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/controller/CourtsController.kt [hunk 6]
```diff
@@ -19,6 +20,8 @@ class CourtsController {
 
   @ResponseStatus(value = HttpStatus.CREATED)
   @PutMapping("/users/{uuid}/preferences/courts")
-  fun updatePreferences(@PathVariable uuid: UUID, @RequestBody user: Users): List<String> =
-    repository.save(Users(uuid, user.courts)).courts
+  fun updatePreferences(@PathVariable uuid: UUID, @RequestBody preference: CourtsDTO): List<String> {
```
Comment (chris-faulkner-mojdt) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/controller/CourtsController.kt:23]: Sorry Paul, not sure why we need "CourtsDTO". I haven't been close to this but I expected generic JSON (like the database table), along the lines of 
{
  user_id : "chris",
  key: "courts",
  values: [
    "B10JQ", 
    "B20XX"
  ]
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/jpa/entity/Users.kt [hunk 7]
```diff
@@ -1,17 +1,21 @@
...
 import javax.persistence.Entity
 import javax.persistence.Id
 
 @Entity
-@TypeDef(name = "list-array", typeClass = ListArrayType::class)
-data class Users(
-  @Id val id: UUID,
-  @Type(type = "list-array") @Column(name = "courts", columnDefinition = "text[]")
-  var courts: List<String>
+@TypeDef(name = "hstore", typeClass = PostgreSQLHStoreType::class)
```
Comment (chris-faulkner-mojdt) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsuserpreferences/jpa/entity/Users.kt:13]: never used this type before in Postgres but looks like just the job.

**Comment by chris-faulkner-mojdt** [comment]
In general, we also have spring boot integration tests which use H2 as the database. We have them in court case service and community API. I think we ought to have them here too.

**Comment by jhevans** [comment]
Closing as this PR has been superseded

**Pull request closed without merging**